File: android/app/src/main/java/com/workflow/tracker/WorkflowBillingPlugin.java
package com.workflow.tracker;

import android.app.Activity;

import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.ConsumeParams;
import com.android.billingclient.api.ProductDetails;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.PurchasesUpdatedListener;
import com.android.billingclient.api.QueryProductDetailsParams;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.Collections;
import java.util.List;

@CapacitorPlugin(name = "WorkflowBilling")
public class WorkflowBillingPlugin extends Plugin implements PurchasesUpdatedListener {
    private BillingClient billingClient;
    private PluginCall pendingCall;

    @PluginMethod
    public void purchase(PluginCall call) {
        String sku = call.getString("sku");
        if (sku == null || sku.trim().isEmpty()) {
            call.reject("Missing sku");
            return;
        }

        if (pendingCall != null) {
            call.reject("Another purchase is already in progress.");
            return;
        }

        pendingCall = call;
        String safeSku = sku.trim();
        ensureClient(() -> queryProductAndLaunch(safeSku));
    }

    private void ensureClient(Runnable onReady) {
        BillingClient client = billingClient;
        if (client == null) {
            client = BillingClient.newBuilder(getContext())
                .setListener(this)
                .enablePendingPurchases()
                .build();
            billingClient = client;
        }

        if (client.isReady()) {
            onReady.run();
            return;
        }

        client.startConnection(new BillingClientStateListener() {
            @Override
            public void onBillingServiceDisconnected() {
                rejectPending("Billing service disconnected.");
            }

            @Override
            public void onBillingSetupFinished(BillingResult result) {
                if (result.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                    onReady.run();
                    return;
                }
                rejectPending("Billing unavailable: " + result.getDebugMessage());
            }
        });
    }

    private void queryProductAndLaunch(String sku) {
        BillingClient client = billingClient;
        if (client == null) {
            rejectPending("Billing unavailable.");
            return;
        }

        QueryProductDetailsParams.Product product =
            QueryProductDetailsParams.Product.newBuilder()
                .setProductId(sku)
                .setProductType(BillingClient.ProductType.INAPP)
                .build();

        QueryProductDetailsParams params =
            QueryProductDetailsParams.newBuilder()
                .setProductList(Collections.singletonList(product))
                .build();

        client.queryProductDetailsAsync(params, (result, products) -> {
            if (result.getResponseCode() != BillingClient.BillingResponseCode.OK
                || products == null
                || products.isEmpty()) {
                rejectPending("Unable to load product details.");
                return;
            }

            ProductDetails details = products.get(0);
            BillingFlowParams.ProductDetailsParams productParams =
                BillingFlowParams.ProductDetailsParams.newBuilder()
                    .setProductDetails(details)
                    .build();

            BillingFlowParams flowParams =
                BillingFlowParams.newBuilder()
                    .setProductDetailsParamsList(Collections.singletonList(productParams))
                    .build();

            Activity currentActivity = getActivity();
            if (currentActivity == null) {
                rejectPending("Unable to launch billing flow.");
                return;
            }

            BillingResult launch = client.launchBillingFlow(currentActivity, flowParams);
            if (launch.getResponseCode() != BillingClient.BillingResponseCode.OK) {
                rejectPending("Billing launch failed: " + launch.getDebugMessage());
            }
        });
    }

    @Override
    protected void handleOnDestroy() {
        super.handleOnDestroy();
        if (billingClient != null) {
            billingClient.endConnection();
            billingClient = null;
        }
        pendingCall = null;
    }

    @Override
    public void onPurchasesUpdated(BillingResult result, List<Purchase> purchases) {
        PluginCall call = pendingCall;
        if (call == null) return;

        int responseCode = result.getResponseCode();
        if (responseCode == BillingClient.BillingResponseCode.OK) {
            Purchase purchase = purchases != null && !purchases.isEmpty() ? purchases.get(0) : null;
            if (purchase == null) {
                rejectPending("No purchase returned.");
                return;
            }
            handlePurchase(purchase);
            return;
        }

        if (responseCode == BillingClient.BillingResponseCode.USER_CANCELED) {
            rejectPending("Purchase canceled.");
            return;
        }

        rejectPending("Purchase failed: " + result.getDebugMessage());
    }

    private void handlePurchase(Purchase purchase) {
        if (purchase.getPurchaseState() != Purchase.PurchaseState.PURCHASED) {
            rejectPending("Purchase not completed.");
            return;
        }

        BillingClient client = billingClient;
        if (client == null) {
            rejectPending("Billing unavailable.");
            return;
        }

        ConsumeParams consumeParams =
            ConsumeParams.newBuilder()
                .setPurchaseToken(purchase.getPurchaseToken())
                .build();

        client.consumeAsync(consumeParams, (result, token) -> {
            if (result.getResponseCode() == BillingClient.BillingResponseCode.OK) {
                JSObject data = new JSObject();
                data.put("ok", true);
                data.put("token", token);
                resolvePending(data);
                return;
            }
            rejectPending("Unable to finalize purchase: " + result.getDebugMessage());
        });
    }

    private void rejectPending(String message) {
        PluginCall call = pendingCall;
        pendingCall = null;
        if (call != null) call.reject(message);
    }

    private void resolvePending(JSObject data) {
        PluginCall call = pendingCall;
        pendingCall = null;
        if (call != null) call.resolve(data);
    }
}

File: native/android/WorkflowBillingPlugin.kt
package com.workflow.tracker

import android.app.Activity
import com.getcapacitor.*
import com.android.billingclient.api.*

@CapacitorPlugin(name = "WorkflowBilling")
class WorkflowBillingPlugin : Plugin(), PurchasesUpdatedListener {
  private var billingClient: BillingClient? = null
  private var pendingCall: PluginCall? = null

  @PluginMethod
  fun purchase(call: PluginCall) {
    val sku = call.getString("sku")
    if (sku.isNullOrBlank()) {
      call.reject("Missing sku")
      return
    }

    if (pendingCall != null) {
      call.reject("Another purchase is already in progress.")
      return
    }

    pendingCall = call
    ensureClient(call) {
      queryProductAndLaunch(sku, call)
    }
  }

  private fun ensureClient(call: PluginCall, onReady: () -> Unit) {
    val client = billingClient ?: BillingClient.newBuilder(context)
      .setListener(this)
      .enablePendingPurchases()
      .build()
      .also { billingClient = it }

    if (client.isReady) {
      onReady()
      return
    }

    client.startConnection(object : BillingClientStateListener {
      override fun onBillingServiceDisconnected() {
        pendingCall?.reject("Billing service disconnected.")
        pendingCall = null
      }

      override fun onBillingSetupFinished(result: BillingResult) {
        if (result.responseCode == BillingClient.BillingResponseCode.OK) {
          onReady()
          return
        }
        pendingCall?.reject("Billing unavailable: ${result.debugMessage}")
        pendingCall = null
      }
    })
  }

  private fun queryProductAndLaunch(sku: String, call: PluginCall) {
    val product = QueryProductDetailsParams.Product.newBuilder()
      .setProductId(sku)
      .setProductType(BillingClient.ProductType.INAPP)
      .build()

    val params = QueryProductDetailsParams.newBuilder()
      .setProductList(listOf(product))
      .build()

    billingClient?.queryProductDetailsAsync(params) { result, products ->
      if (result.responseCode != BillingClient.BillingResponseCode.OK || products.isEmpty()) {
        call.reject("Unable to load product details.")
        pendingCall = null
        return@queryProductDetailsAsync
      }

      val details = products.first()
      val productParams = BillingFlowParams.ProductDetailsParams.newBuilder()
        .setProductDetails(details)
        .build()

      val flowParams = BillingFlowParams.newBuilder()
        .setProductDetailsParamsList(listOf(productParams))
        .build()

      val currentActivity: Activity? = activity
      if (currentActivity == null) {
        call.reject("Unable to launch billing flow.")
        pendingCall = null
        return@queryProductDetailsAsync
      }

      val launch = billingClient?.launchBillingFlow(currentActivity, flowParams)
      if (launch?.responseCode != BillingClient.BillingResponseCode.OK) {
        call.reject("Billing launch failed: ${launch?.debugMessage}")
        pendingCall = null
      }
    }
  }

  override fun onPurchasesUpdated(result: BillingResult, purchases: MutableList<Purchase>?) {
    val call = pendingCall ?: return

    when (result.responseCode) {
      BillingClient.BillingResponseCode.OK -> {
        val purchase = purchases?.firstOrNull()
        if (purchase == null) {
          call.reject("No purchase returned.")
          pendingCall = null
          return
        }
        handlePurchase(call, purchase)
      }
      BillingClient.BillingResponseCode.USER_CANCELED -> {
        call.reject("Purchase canceled.")
        pendingCall = null
      }
      else -> {
        call.reject("Purchase failed: ${result.debugMessage}")
        pendingCall = null
      }
    }
  }

  private fun handlePurchase(call: PluginCall, purchase: Purchase) {
    if (purchase.purchaseState != Purchase.PurchaseState.PURCHASED) {
      call.reject("Purchase not completed.")
      pendingCall = null
      return
    }

    val consumeParams = ConsumeParams.newBuilder()
      .setPurchaseToken(purchase.purchaseToken)
      .build()

    billingClient?.consumeAsync(consumeParams) { result, token ->
      if (result.responseCode == BillingClient.BillingResponseCode.OK) {
        val data = JSObject()
        data.put("ok", true)
        data.put("token", token)
        call.resolve(data)
      } else {
        call.reject("Unable to finalize purchase: ${result.debugMessage}")
      }
      pendingCall = null
    }
  }
}
